### ISSUE TITLE
Mixin pattern introduced in 2.2 fails when type declaration order is screwed

### ISSUE THREAD

**Issue by Dynalon** [Microsoft/TypeScript#15316]
**TypeScript Version:**  2.2.2 

The following code examples compiles successfully (default compiler options, no tsconfig) but will give a runtime error: `TypeError: Object prototype may only be an Object or null: undefined` when run with node.

When applying the mixin after the declaration of the `Person` type, the code runs as expected. While this is somewhat reasonable, I would expect a compile error or at least a warning when compiling with TypeScript that the `Person` constructor is undefined at this point.

**Code**

```ts
type Constructor<T> = new(...args: any[]) => T;

function Mixin<T extends Constructor<{}>>(Base: T) {

    return class extends Base {
        age: number = 99;

        constructor(...args: any[]) {
            super(...args);
        }
    }
}

// does not work when placing here:
const OldPerson = Mixin(Person);

class Person {
    constructor(public name: string) {
    }
}

// when declared after Person, this will work:
// const OldPerson = Mixin(Person);

let old = new OldPerson("John");
console.log(old.age);
```

**Expected behavior:**

Correctly apply Mixin **or** give compilation error/warning when compiling.

**Actual behavior:**

Successfull compile but runtime error

**Comment by mhegazy** [comment]
Should be reporting the error correctly in latest ([playground link](http://www.typescriptlang.org/play/index.html#src=type%20Constructor%3CT%3E%20%3D%20new(...args%3A%20any%5B%5D)%20%3D%3E%20T%3B%0D%0A%0D%0Afunction%20Mixin%3CT%20extends%20Constructor%3C%7B%7D%3E%3E(Base%3A%20T)%20%7B%0D%0A%0D%0A%20%20%20%20return%20class%20extends%20Base%20%7B%0D%0A%20%20%20%20%20%20%20%20age%3A%20number%20%3D%2099%3B%0D%0A%0D%0A%20%20%20%20%20%20%20%20constructor(...args%3A%20any%5B%5D)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20super(...args)%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20does%20not%20work%20when%20placing%20here%3A%0D%0Aconst%20OldPerson%20%3D%20Mixin(Person)%3B%0D%0A%0D%0Aclass%20Person%20%7B%0D%0A%20%20%20%20constructor(public%20name%3A%20string)%20%7B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20when%20declared%20after%20Person%2C%20this%20will%20work%3A%0D%0A%2F%2F%20const%20OldPerson%20%3D%20Mixin(Person)%3B%0D%0A%0D%0Alet%20old%20%3D%20new%20OldPerson(%22John%22)%3B%0D%0Aconsole.log(old.age)%3B))
